lib/icinga/apiactions.cpp

```diff
@@ -555,5 +555,179 @@ Dictionary::Ptr ApiActions::GenerateTicket(const ConfigObject::Ptr&,
 Dictionary::Ptr ApiActions::ExecuteCommand(const ConfigObject::Ptr& object,
 	const Dictionary::Ptr& params)
 {
-	return ApiActions::CreateResult(501, "Not implemented");
+	Checkable::Ptr checkable = static_pointer_cast<Checkable>(object);
+	if (! checkable)
+		return ApiActions::CreateResult(404, "Can't schedule a command execution for non-existent object.");
+
+	ObjectLock oLock (checkable);
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
+
+	String endpoint = "$command_endpoint$";
+	if (params->Contains("endpoint"))
+		endpoint = params->Get("endpoint");
+
+	/* Resolve endpoint macro */
+	MacroProcessor::ResolverList resolvers;
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("icinga", IcingaApplication::GetInstance());
+
+	Dictionary::Ptr resolvedMacros;
+	bool useResolvedMacros;
+
+	if (params->Contains("macros")) {
+		resolvedMacros = HttpUtility::GetLastParameter(params, "macros");
```

Comment:
What happens if you specify `curl -d '{"macros": "a string, not a dictionary"}'`?